### PR TITLE
[SPARK-24392][PYTHON] Label pandas_udf as Experimental

### PR DESCRIPTION
**Pull request by BryanCutler**
## What changes were proposed in this pull request?

The pandas_udf functionality was introduced in 2.3.0, but is not completely stable and still evolving.  This adds a label to indicate it is still an experimental API.

## How was this patch tested?

NA

### CODE REVIEW

**Comment by BryanCutler** [comment]
@gatorsmile @HyukjinKwon , I think we should probably make this change in light of some the recent issues, but I understand if we can not because it has already been released.  What are your thoughts?

**Comment by HyukjinKwon** [comment]
I think it's just a mistake. It should have been done eariler. I think we should restore this righer description for this new feature before it's too late. It was a big new feature and still incomplete.

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -2448,6 +2448,8 @@ def udf(f=None, returnType=StringType()):
 @since(2.3)
 def pandas_udf(f=None, returnType=None, functionType=None):
     """
+    .. note:: Experimental
+
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: @BryanCutler, can we add a short node for createDataFrame and to Pandas with Pandas DataFrame input/output?

Comment (BryanCutler) [reply python/pyspark/sql/functions.py]: Yeah, I could have sworn that was added before but I don't see it anywhere.  Would it make sense to add the note to the SQLconf description?

Comment (HyukjinKwon) [reply python/pyspark/sql/functions.py]: Anywhere is fine IMHO if it looks making sense. Another option is  things like: 

```
.. versionchanged:: 2.3
  Arrow optimization added
  .. note:: Experimental. 
```

I just wanted to make sure this feature is experimental. If it's wrongly designed, we should consider a behaviour change as the last resort.

Comment (HyukjinKwon) [reply python/pyspark/sql/functions.py]: BTW, don't forget to update `group.apply()`.

Comment (BryanCutler) [reply python/pyspark/sql/functions.py]: Trying to add the note under `versionchanged` ends up with everything on one line and the note isn't highlighted like usual with Experimental, so I just made it it's own note.

**Review by HyukjinKwon — APPROVED** [review]
LGTM

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -2448,6 +2448,8 @@ def udf(f=None, returnType=StringType()):
 @since(2.3)
 def pandas_udf(f=None, returnType=None, functionType=None):
     """
+    .. note:: Experimental
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: not a big deal but can we put this right before doctest just for consistency?

**Comment by viirya** [comment]
I think this LGTM, should we add a note in migration doc from 2.3.0 to 2.3.1 for this change?

**Comment by HyukjinKwon** [comment]
yup Im seeing no harn to mtnetion it. It shouldn't be used in production for now.

File: python/pyspark/sql/functions.py (hunk 2 above)
Comment (BryanCutler) [reply python/pyspark/sql/functions.py]: yeah, it was hard to find a consistent place to put it.  I changed this one to be after the param definitions, before doctest.  hopefully that's better.

**Comment by BryanCutler** [comment]
I added a note in the migration guide for 2.3 to 2.4, but in branch-2.3 I don't see a section for 2.3.0 to 2.3.1.  When backporting should I make that section for this or ok to just place at the end of 2.3?

**Comment by HyukjinKwon** [comment]
oh yea please.could we just make a section "Upgrading From Spark SQL 2.3.0 to 2.3.1 and above " here?

**Comment by viirya** [comment]
@HyukjinKwon's suggestion sounds good to me.

**Comment by BryanCutler** [comment]
Ahh, good idea thanks!

File: docs/sql-programming-guide.md [hunk 3]
```diff
@@ -1827,6 +1827,10 @@ working with timestamps in `pandas_udf`s to get the best performance, see
...
   - In version 2.3 and earlier, CSV rows are considered as malformed if at least one column value in the row is malformed. CSV parser dropped such rows in the DROPMALFORMED mode or outputs an error in the FAILFAST mode. Since Spark 2.4, CSV row is considered as malformed only when it contains malformed column values requested from CSV datasource, other values can be ignored. As an example, CSV file contains the "id,name" header and one row "1234". In Spark 2.4, selection of the id column consists of a row with one column value 1234 but in Spark 2.3 and earlier it is empty in the DROPMALFORMED mode. To restore the previous behavior, set `spark.sql.csv.parser.columnPruning.enabled` to `false`.
 
+## Upgrading From Spark SQL 2.3.0 to 2.3.1 and Above
```
Comment (HyukjinKwon) [inline docs/sql-programming-guide.md]: sorry `Above` -> `above`?

Comment (BryanCutler) [reply docs/sql-programming-guide.md]: sure

**Comment by HyukjinKwon** [comment]
Thanks @BryanCutler. still LGTM

**Comment by HyukjinKwon** [comment]
Merged to master and branch-2.3.

There were minor conflicts but I just resolved it by myself.

**Pull request closed without merging**